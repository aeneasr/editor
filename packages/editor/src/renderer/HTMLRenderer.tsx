import classNames from 'classnames';
import React from 'react';
import { migrateValue } from '../core/migrations/migrate';
import { setAllSizesAndOptimize } from '../core/reducer/value/helper/setAllSizesAndOptimize';
import { optimizeRows } from '../core/reducer/value/helper/optimize';
import type { Cell, CellPlugin, Row, ValueWithLegacy } from '../core/types';
import { getCellData } from '../core/utils/getCellData';
import { getCellStyle } from '../core/utils/getCellStyle';
import {
  getPluginCellSpacing,
  normalizeCellSpacing,
} from '../core/utils/getCellSpacing';

const gridClass = (size = 12): string =>
  `react-page-cell-sm-${size} react-page-cell-xs-12`;

const rowHasInlineChildren = ({ cells }) =>
  Boolean(cells.length === 2 && Boolean(cells[0].inline));

const HTMLRow: React.FC<
  Partial<
    Row & {
      lang: string;
      className?: string;
      cellPlugins: CellPlugin[];
      cellSpacing: [number, number];
    }
  >
> = React.memo(
  ({
    cells = [],
    className,
    lang,
    cellPlugins,
    cellSpacing: [cellSpacingX, cellSpacingY],
  }) => (
    <div
      className={classNames('react-page-row', className, {
        'react-page-row-has-floating-children': rowHasInlineChildren({ cells }),
      })}
      style={{ margin: `0 ${-cellSpacingX / 2}px` }}
    >
      {cells.map((c, index) => (
        <HTMLCell
          key={c.id}
          {...c}
          lang={lang}
          cellPlugins={cellPlugins}
          cellSpacing={[cellSpacingX, cellSpacingY]}
        />
      ))}
    </div>
  )
);

// eslint-disable-next-line no-empty-function
const noop = () => {
  return;
};

const DefaultProvider: React.FC = ({ children }) => <>{children}</>;

const HTMLCell: React.FC<
  Cell & {
    lang: string;
    cellPlugins: CellPlugin[];
    cellSpacing: [number, number];
  }
> = React.memo((props) => {
  const {
    lang,
    cellPlugins,
    cellSpacing: [cellSpacingX, cellSpacingY],
    ...cell
  } = props;
  const { size, hasInlineNeighbour, inline, isDraftI18n, isDraft } = cell;
  const hasChildren = cell.rows?.length > 0;
  const cnOuter = classNames(gridClass(size), {
    'react-page-cell-has-inline-neighbour': hasInlineNeighbour,
    [`react-page-cell-inline-${inline || ''}`]: inline,
  });
  const cnInner = classNames('react-page-cell', {
    'react-page-cell-leaf': !hasChildren,
  });

  if (isDraftI18n?.[lang] ?? isDraft) {
    return null;
  }

  const plugin = cell.plugin
    ? cellPlugins.find((p) => p.id === cell.plugin.id)
    : null;
  if (plugin) {
    const { Renderer } = plugin;
    const cellStyle = getCellStyle(plugin);
    const Provider = plugin.Provider ?? DefaultProvider;
    const data = getCellData(cell, lang);
    const pluginCellSpacing = getPluginCellSpacing(plugin, data);
    const [pluginCellSpacingX, pluginCellSpacingY] = pluginCellSpacing
      ? normalizeCellSpacing(pluginCellSpacing)
      : [cellSpacingX, cellSpacingY];
    const props = {
      readOnly: true,
      lang: lang,
      nodeId: cell.id,
      data: data,
      onChange: noop,
      pluginConfig: plugin,
      focused: false,
      isPreviewMode: false,
      isEditMode: false,
    };
    return (
      <Provider {...props}>
        <div
          className={cnOuter}
          style={{ padding: `${cellSpacingY / 2}px ${cellSpacingX / 2}px` }}
        >
          <div className={cnInner}>
            <div
              style={cellStyle}
              className={
                'react-page-cell-inner' +
                (cell.rows?.length > 0 ? '' : ' react-page-cell-inner-leaf')
              }
            >
              <div
                style={{
                  display: 'flex',
                  flexDirection: 'column',
                }}
              >
                <Renderer {...props}>
                  {cell.rows?.length ? (
                    <div style={{ margin: `${-pluginCellSpacingY / 2}px 0` }}>
                      {cell.rows?.map((r: Row) => (
                        <HTMLRow
                          key={r.id}
                          {...r}
                          cellPlugins={cellPlugins}
                          cellSpacing={[pluginCellSpacingX, pluginCellSpacingY]}
                          lang={lang}
                          className="react-page-cell-inner"
                        />
                      ))}
                    </div>
                  ) : null}
                </Renderer>
              </div>
            </div>
          </div>
        </div>
      </Provider>
    );
  } else if (cell.rows?.length > 0) {
    return (
      <div className={cnOuter} style={{ padding: `0 ${cellSpacingX / 2}px` }}>
        <div className={cnInner}>
          {cell.rows.map((r: Row) => (
            <HTMLRow
              key={r.id}
              {...r}
              lang={lang}
              className="react-page-cell-inner"
              cellPlugins={cellPlugins}
              cellSpacing={[cellSpacingX, cellSpacingY]}
            />
          ))}
        </div>
      </div>
    );
  }

  return (
    <div className={cnOuter}>
      <div className={cnInner}>
        <div className="react-page-cell-inner" />
      </div>
    </div>
  );
});

export interface HTMLRendererProps {
  value: ValueWithLegacy;
  cellPlugins?: CellPlugin[];
  lang?: string;
  cellSpacing?: number | [number, number];
}

export const HTMLRenderer: React.FC<HTMLRendererProps> = React.memo(
  ({ value, cellPlugins, cellSpacing, lang = 'default' }) => {
    const data = migrateValue(value, { cellPlugins, lang });
    const [cellSpacingX, cellSpacingY] = normalizeCellSpacing(cellSpacing);

    if (!data) {
      return null;
    }
    const { rows } = data;
    const optRows = optimizeRows(rows);
    return (
      <div
        style={{ margin: optRows?.length ? `${-cellSpacingY / 2}px 0` : null }}
      >
        {setAllSizesAndOptimize(optRows).map((row) => (
          <HTMLRow
            key={row.id}
            cellPlugins={cellPlugins}
            lang={lang}
            cellSpacing={[cellSpacingX, cellSpacingY]}
            {...row}
          />
        ))}
      </div>
    );
  }
);
