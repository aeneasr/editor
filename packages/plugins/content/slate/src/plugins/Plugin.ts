/*
 * This file is part of ORY Editor.
 *
 * ORY Editor is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * ORY Editor is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with ORY Editor.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @license LGPL-3.0
 * @copyright 2016-2018 Aeneas Rekkas
 * @author Aeneas Rekkas <aeneas+oss@aeneas.io>
 *
 */
import { RenderMarkProps, RenderNodeProps } from 'slate-react';
import { Value } from 'slate';

export class PluginButtonProps {
  editorState: Value;
  focus: boolean;
  onChange: (state: { value: Value }) => void;
}

/**
 * @class this is the base class for slate plugins
 */
export default class Plugin {
  /**
   * @member a default node
   */
  public DEFAULT_NODE: string;

  /**
   * @member a unique identifier of the plugin
   */
  public name: string;

  /**
   * @member the schema that is automatically collected from all plugins
   */
  schema: {
    nodes?: { [key: string]: React.SFC };
    marks?: { [key: string]: React.SFC };
  };

  /**
   * @member the slate plugins added to the editor
   */
  plugins: Plugin[] = [];

  /**
   * @member serialize a plugin's state to html
   */
  serialize: (
    // tslint:disable-next-line:no-any
    object: { object: string; type: string; data: any },
    // tslint:disable-next-line:no-any
    children: any[]
  ) => // tslint:disable-next-line:no-any
  any;

  /**
   * @member serialize a plugin's state from html
   */
  // tslint:disable-next-line:no-any
  deserialize: (el: Element, next: Function) => any;

  /**
   * @member the buttons to be added to the hover menu
   */
  public hoverButtons: (
    | React.ComponentClass<PluginButtonProps>
    | React.SFC<PluginButtonProps>)[];

  /**
   * @member the buttons to be added to the global toolbar
   */
  public toolbarButtons: (
    | React.ComponentClass<PluginButtonProps>
    | React.SFC<PluginButtonProps>)[];

  /**
   * @member the function that renders marks
   */
  public renderMark: (props: RenderMarkProps) => JSX.Element;

  /**
   * @member the function that renders nodes
   */
  public renderNode: (props: RenderNodeProps) => JSX.Element;

  /**
   * This handler is called when any key is pressed
   *
   * @param e the keydown event
   * @param data utilities for hotkey logic
   * @param state the current editor state
   * @returns the new editor state if the plugin handles the hotkey
   */
  public onKeyDown = (
    e: Event,
    data: { key: string; isMod: boolean; isShift: boolean },
    // tslint:disable-next-line:no-any
    state: any
    // tslint:disable-next-line:no-any
  ): any => null
}
