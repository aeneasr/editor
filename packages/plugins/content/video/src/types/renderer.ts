import type { VideoProps } from './component';

export type VideoHtmlRendererProps = VideoProps;
