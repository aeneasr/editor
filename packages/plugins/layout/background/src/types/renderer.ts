import { BackgroundProps } from './component';
import { RGBColor } from 'ory-editor-ui/lib/ColorPicker';
import { ImageLoaded } from 'ory-editor-ui/lib/ImageUpload';

export interface BackgroundRendererExtraProps {
  backgroundColorPreview?: RGBColor;
  gradientDegPreview?: number;
  gradientDegPreviewIndex?: number;
  gradientOpacityPreview?: number;
  gradientOpacityPreviewIndex?: number;
  gradientColorPreview?: RGBColor;
  gradientColorPreviewIndex?: number;
  gradientColorPreviewColorIndex?: number;
  darkenPreview?: number;
  lightenPreview?: number;
  imagePreview?: ImageLoaded;
  forecolorPreview?: string;
  backgroundSize?: string;
  backgroundPositionH?: string;
  backgroundPositionV?: string;
}

export type BackgroundRendererProps = BackgroundProps & BackgroundRendererExtraProps;