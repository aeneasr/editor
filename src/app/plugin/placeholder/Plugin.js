import AbstractPlugin from 'app/service/plugin/Plugin';
import React from 'react';
import map from 'lodash/collection/map';
import Placeholder from './Component';

export default class PlaceholderPlugin extends AbstractPlugin {
    constructor() {
        super({
            name: 'placeholder',
            version: 0
        });
    }

    render(sections, store) {
        /*jshint ignore:start */
        return <div>{map(sections, () => <Placeholder />)}</div>;
        /*jshint ignore:end */
    }
}
