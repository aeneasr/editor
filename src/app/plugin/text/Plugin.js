import AbstractPlugin from 'app/service/plugin/Plugin';
import Sections from './Component';
import React from 'react';

export default class TextPlugin extends AbstractPlugin {
    constructor() {
        super({name: 'text', version: 0});
    }

    render(sections) {
        /*jshint ignore:start */
        return <Sections sections={sections}></Sections>;
        /*jshint ignore:end */
    }
}
