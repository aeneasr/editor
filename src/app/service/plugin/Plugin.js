import Section from 'app/entity/Section';
import InvalidArgumentException from 'app/exception/InvalidArgumentException';
import {Component} from 'react';
import React from 'react';
import Actions from 'app/Actions';

export default class Plugin {
    constructor(options) {
        var reg = /([0-9]\.)+[0-9]/i;

        if (options.name === undefined || options.length === 0) {
            throw new InvalidArgumentException('options.name', 'String (not empty)', options.name);
        }

        if (!options.version.match(reg)) {
            throw new InvalidArgumentException('options.version', 'String (should match `[X.]+X`)', options.version);
        }

        if (typeof options.section !== 'function') {
            throw new InvalidArgumentException('options.section', 'function', options.section);
        }

        if (typeof options.toolbar !== 'function') {
            options.toolbar = Toolbar;
        }

        // This is the plugin function (should be a React Component)
        this.Section = options.section;
        this.Toolbar = options.toolbar;

        // The plugin's settings
        this.name = options.name || '';
        this.version = options.version || '';
        this.compatibility = options.compatibility || [];
        this.args = options.args || {};

        // Some callbacks for easy extension of this class
        this.toolbarcb = options.toolbar;
        this.parsecb = options.parse;
        this.createcb = options.create;
        this.join = options.join || false;
    }

    focus(context) {
        Actions.editable.section.focus(this, context);
    }

    blur(context) {
        Actions.editable.section.blur(this, context);
    }

    parse(element, options) {
        if (typeof this.parsecb === 'function') {
            return this.parsecb(element, options);
        }
        return {innerHTML: element.innerHTML};
    }

    createSection() {
        if (typeof this.createcb === 'function') {
            return this.createcb();
        }
        return new Section(this.name, this.version, this.args, {innerHTML: '<p>Click in this area to write text</p>'});
    }
}

