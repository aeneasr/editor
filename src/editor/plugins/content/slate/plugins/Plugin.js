// @flow
/* eslint-disable no-unused-vars */
import { Component } from 'react'
// import { State } from 'slate'

/**
 * @class this is the base class for slate plugins
 */
export default class Plugin {
  /**
   * @member a unique identifier of the plugin
   */
  name: string

  /**
  * @member the nodes to be added to the schema
  */
  nodes: { [key: string]: Component<*, *, *> } = {}

  /**
   * @member the marks to be added to the schema
   */
  marks: { [key: string]: Component<*, *, *> } = {}

  /**
   * @member the slate plugins added to the editor
   */
  plugins: Array<*> = []

  /**
   * This handler is called when any key is pressed
   *
   * @param e the keydown event
   * @param data utilities for hotkey logic
   * @param state the current editor state
   * @returns the new editor state if the plugin handles the hotkey
   */
  onKeyDown = (
    e: Event,
    data: { key: string, isMod: bool, isShift: bool },
    state: any
  ): ?any => null

  /**
   * @member the buttons to be added to the hover menu
   */
  hoverButtons: Array<Component<*, *, *>> = []

  /**
   * @member the buttons to be added to the global toolbar
   */
  toolbarButtons: Array<Component<*, *, *>> = []
}
