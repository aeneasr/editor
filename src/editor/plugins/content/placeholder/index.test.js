/* eslint-env jest */
import React from 'react'
import renderer from 'react-test-renderer'

import Plugin from './index'

jest.mock('react-dom')

describe('plugins/content/placeholder/Component', () => {
  it('should render', () => {
    const { Component } = new Plugin()
    const component = renderer.create(<Component />)
    expect(component).toMatchSnapshot()
  })
})
