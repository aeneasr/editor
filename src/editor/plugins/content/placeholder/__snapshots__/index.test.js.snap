exports[`plugins/content/placeholder/Component should render 1`] = `
<div
  className="placeholder">
  Drop items above, below, left of or right of here.
</div>
`;
